### ISSUE TITLE
Masks shape doesn't match input shape if retina_masks=False

### ISSUE THREAD

**Issue by charlielito** [ultralytics/ultralytics#8691]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

The masks tensor doesn't match the shape of the input image if `retina_masks=False`. When  `retina_masks=True` which is not the default, the shapes are OK.

### Environment

Ultralytics YOLOv8.1.23 🚀 Python-3.8.18 torch-2.2.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3070 Laptop GPU, 7974MiB)
Setup complete ✅ (8 CPUs, 15.3 GB RAM, 697.8/778.0 GB disk)

OS                  Linux-6.0.12-76060012-generic-x86_64-with-glibc2.17
Environment         Linux
Python              3.8.18
Install             git
RAM                 15.32 GB
CPU                 11th Gen Intel Core(TM) i7-11370H 3.30GHz
CUDA                12.1

matplotlib          ✅ 3.7.1>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 9.5.0>=7.1.2
pyyaml              ✅ 6.0>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.10.0>=1.4.1
torch               ✅ 2.2.0>=1.8.0
torchvision         ✅ 0.17.0>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.5
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0

### Minimal Reproducible Example

![car](https://github.com/ultralytics/ultralytics/assets/8033598/97807e49-49b9-46fe-8c69-585848e37eaf)

```python
from ultralytics import YOLO

model = YOLO("yolov8s-seg.pt")

result = model("car.png", verbose=False)[0]
masks_tensor = result.masks.cpu().data
print("Shapes with defaults retina_masks=False")
print(f"Shape of input image: {result.orig_shape}")
print(f"Shape of masks_tensor: {masks_tensor.shape}")

print("\nShapes with retina_masks=True")
result = model("car.png", verbose=False, retina_masks=True)[0]
masks_tensor = result.masks.cpu().data
print(f"Shape of input image: {result.orig_shape}")
print(f"Shape of masks_tensor: {masks_tensor.shape}")

```

Output:

```
Shapes with defaults retina_masks=False
Shape of input image: (776, 1300)
Shape of masks_tensor: torch.Size([2, 384, 640])

Shapes with retina_masks=True
Shape of input image: (776, 1300)
Shape of masks_tensor: torch.Size([2, 776, 1300])
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@charlielito masks I think should be shaped to imgsz, which is 640. Retina masks evaluates masks at higher resolution, which is returning the larger image size. 

@Laughing-q what do you think here?

**Comment by charlielito** [comment]
@glenn-jocher with retina mask the mask has always the shape of the input image, as **it should** because even the boxes are returned with respect to the input image shape. IMO is very confusing to the user that sometimes it is returned with respect to the original input dimensions and sometimes with that internal hardcoded 640

**Comment by glenn-jocher** [comment]
Hi @charlielito, thanks for bringing this to our attention! 🌟 You're absolutely right; consistency in the output shapes, especially between masks and bounding boxes, is crucial for a seamless user experience. The behavior you've observed with `retina_masks` indeed aligns with the expectation that masks should match the input image dimensions, similar to how bounding boxes are handled.

We'll look into making the output more intuitive and consistent, ensuring that both masks and bounding boxes align with the input image dimensions, regardless of the internal processing size. This should help eliminate any confusion and make the API more user-friendly.

In the meantime, if you need to resize the masks to match your input image dimensions when `retina_masks=False`, you can use the following workaround:

```python
import cv2
import torch

# Assuming masks_tensor is your output mask tensor and orig_shape is the shape of your input image
resized_masks = torch.nn.functional.interpolate(masks_tensor.unsqueeze(0), size=orig_shape[::-1], mode='bilinear', align_corners=False).squeeze(0)
```

This snippet uses bilinear interpolation to resize the masks to match your original input image dimensions. 

Thanks again for your feedback, and please stay tuned for updates on this issue!

**Comment by Laughing-q** [comment]
@charlielito @glenn-jocher Hey guys! we designed `retina_masks` to fit the situations that sometimes users want faster speed and less memory, sometimes users want more accurate masks.
- `retina_masks=False` would return masks with `imgsz` which usually is 640 and smaller than original resolution, and it comes with faster plot speed and less memory occupation.
- `retina_masks=True` returns masks with original resolution to fit the original objects better, the masks are more accurate.
